tests/e2e/test_step_types.py

```diff
@@ -0,0 +1,80 @@
+import subprocess
+import sys
+import tempfile
+from pathlib import Path
+
+import pytest
+
+from tests.conftest import RESULTS_DIR, SPECIFICATIONS_DIR
+
+
+@pytest.mark.slow
+@pytest.mark.parametrize(
+    "pipeline_specification, implementations",
+    [
+        # Basic
+        (
+            "common/pipeline.yaml",
+            [
+                "step_1_python_pandas",
+                "step_2_python_pandas",
+                "step_3_python_pandas",
+                "step_4_python_pandas",
+            ],
+        ),
+        # Substeps
+        (
+            "e2e/pipeline_hierarchical.yaml",
```

Comment:
Kinda what I said in my other comment, but do we even need the `common/pipeline` test anymore? Isn't the hierarchical one testing all functionality of the common one plus some?

Reply:
I think the distinction is that we want to make sure that the *same* step can be configured in multiple different ways--that is, that we can have either a "one implementation" step 1 or a "substep" step 1, as opposed to having one test where we test substeps on step 1, and then subsequently have a basic implementation for say, step 2 or whatever. "common" ensures that each step can function as a basic step.

That said, maybe we can just have essentially two--one for the basic and one that does all the other variations for the other steps.